src/shady-css-parser/parser.js

```diff
@@ -0,0 +1,331 @@
+import { matcher, tokenType } from './common';
+import { TokenCursor } from './token-cursor';
+
+function AtRule(name, parameters, block) {
+  return { type: 'at-rule', name, parameters, block };
+}
+
+function Comment(value) {
+  return { type: 'comment', value };
+}
+
+function Block(declarations) {
+  return { type: 'block', declarations };
+}
+
+function Selector(combinator, block) {
+  return { type: 'selector', combinator, block };
+}
+
+function Property(name, value) {
+  return { type: 'property', name, value };
+}
+
+function Value(text) {
+  return { type: 'value', text };
+}
+
+function ParserError(index, error, ...artifacts) {
+  return { type: 'parser-error', index, error, artifacts };
+}
+
+function Stylesheet(tokenCursor) {
+  let declarations = [];
+
+  while (tokenCursor.next) {
+    try {
+      declarations.push(parseDeclaration(tokenCursor));
+    } catch (error) {
```

Comment:
What exceptions do you expect to encounter here? Since you don't throw anywhere, any exception would seem to be an error in the parser, in which case I'm not sure a AST with a ParseError node is the correct thing to return.